### PR TITLE
[FLINK-5056] Makes the BucketingSink rescalable.

### PR DESCRIPTION
**Pull request by kl0u**
This PR makes the BucketingSink rescalable, fixes a bug that could lead to deleting
valid data and improves the javadocs of the class. 

In the process of making the sink rescalable, we also stop deleting lingering files upon restoring.
This is to avoid possible race-conditions that can lead to one task deleting files that another
task uses.

This also includes issue https://issues.apache.org/jira/browse/FLINK-5054

R @aljoscha

### CODE REVIEW

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 1]
```diff
@@ -234,7 +251,9 @@
 	 * <p>
 	 * This should only be set to false if using the sink without checkpoints, to not remove
 	 * the files already in the directory.
+	 * <p>
 	 */
+	@Deprecated
 	private boolean cleanupOnOpen = true;
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: this field is unused and can be removed

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 2]
```diff
@@ -426,17 +424,63 @@ public void close() throws Exception {
 		}
 	}
 
+	// TODO: 11/10/16 this should be the close() after
+	// introducing the dispose() method in the RichFunction.
+	// ATTENTION: it is not thoroughly tested yet.
+	public void closeAfterDisposeIsIntroduced() throws Exception {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: You remove this method in a later commit, why add it in the first place?

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 3]
```diff
@@ -481,27 +494,27 @@ public void onProcessingTime(long timestamp) throws Exception {
 	}
 
 	/**
-	 * Checks for inactive buckets, and closes them. This enables in-progress files to be moved to
-	 * the pending state and finalised on the next checkpoint.
+	 * Checks for inactive buckets, and closes them. Inactive are buckets that have not been written to
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: Rewording: "Buckets are considered inactive if they have not been ..."

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 4]
```diff
@@ -570,284 +583,278 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
-		if(fs != null) {
-			Class<?> fsClass = fs.getClass();
-			try {
-				m = fsClass.getMethod("truncate", Path.class, long.class);
-			} catch (NoSuchMethodException ex) {
-				LOG.debug("Truncate not found. Will write a file with suffix '{}' " +
-					" and prefix '{}' to specify how many bytes in a bucket are valid.", validLengthSuffix, validLengthPrefix);
-				return null;
-			}
+		if (this.refTruncate == null) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: Why did you change this method? we can check for null before calling truncate instead.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 5]
```diff
@@ -570,284 +583,278 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
-		this.state = state;
-
-		try {
-			initFileSystem();
-		} catch (IOException e) {
-			LOG.error("Error while creating FileSystem in checkpoint restore.", e);
-			throw new RuntimeException("Error while creating FileSystem in checkpoint restore.", e);
-		}
+	private void handleRestoredBucketState(BucketState<T> bucketState) {
+		// we can clean all the pending files since they were renamed to
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: according to the PR description we no longer delete files at this point since multiple tasks may use the same file. Under which circumstances can this occur?

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 6]
```diff
@@ -570,284 +583,278 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
-	public void restoreState(State<T> state) {
-		this.state = state;
-
-		try {
-			initFileSystem();
-		} catch (IOException e) {
-			LOG.error("Error while creating FileSystem in checkpoint restore.", e);
-			throw new RuntimeException("Error while creating FileSystem in checkpoint restore.", e);
-		}
+	private void handleRestoredBucketState(BucketState<T> bucketState) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: Please move the looping over the buckets into this method; this will clean up the diff tremendously.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 7]
```diff
@@ -570,284 +616,277 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
+
 				synchronized (bucketState.pendingFilesPerCheckpoint) {
-					bucketState.pendingFilesPerCheckpoint.put(checkpointId, bucketState.pendingFiles);
+					bucketState.pendingFilesPerCheckpoint.put(context.getCheckpointId(), bucketState.pendingFiles);
 				}
 				bucketState.pendingFiles = new ArrayList<>();
 			}
+			restoredBucketStates.add(state);
+
+			if (LOG.isDebugEnabled()) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: you don't have to check this.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 8]
```diff
@@ -570,284 +616,277 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
+								Path pendingPath = getPendingPathFor(finalPath);
 
 								fs.rename(pendingPath, finalPath);
 								LOG.debug(
 									"Moving pending file {} to final location having completed checkpoint {}.",
 									pendingPath,
 									pastCheckpointId);
 							}
-							checkpointsToRemove.add(pastCheckpointId);
+							pendingIt.remove();
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: rename to "pendingCheckpoints"

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 9]
```diff
@@ -570,284 +616,277 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
+		return new Path(path.getParent(), validLengthPrefix + path.getName()).suffix(validLengthSuffix);
 	}
 
 	@Override
 	public void notifyCheckpointComplete(long checkpointId) throws Exception {
 		synchronized (state.bucketStates) {
-			Iterator<Map.Entry<String, BucketState<T>>> it = state.bucketStates.entrySet().iterator();
-			while (it.hasNext()) {
-				BucketState<T> bucketState = it.next().getValue();
+			Iterator<Map.Entry<String, BucketState<T>>> stateIt = state.bucketStates.entrySet().iterator();
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: if you really want to rename this i would suggest "bucketStates".

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 10]
```diff
@@ -570,284 +616,277 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
+					LOG.error("Could not delete truncate test file.", e);
+					throw new RuntimeException("Could not delete truncate test file.", e);
+				}
 			}
+			this.refTruncate = m;
+		}
+		return this.refTruncate;
+	}
 
+	private Path getPendingPathFor(Path path) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: these helper methods can be static.

**Comment by zentol** [comment]
I also haven't checked the tests yet.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 11]
```diff
@@ -150,53 +162,59 @@
 
 	/**
 	 * The default maximum size of part files.
-	 *
-	 * 6 times the default block size
+	 * <p>
+	 * By default, {@code 6 X} the default block size.
```
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: This existed from before. Just reformatted it. I can remove it if it is clearer.

**Comment by kl0u** [comment]
Thanks for the first review @zentol . 

On the comment about the diff reduction, you are referring to the removal of the method in the intermediate commits that was not used? If yes, then this is the reason it was removed in the next commit. Either way, the commit about the documentation should be squashed before merging.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 11 above)
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: i think we should either not mention "block size" or use a more common term for whatever it stands for.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 5 above)
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: I do not get the question? You mean when the thing on the comment can occur? The comment just says that we remove it from our current state. 

Now if you mean when two tasks are responsible for the same files, then consider the  following: 

Files written by a task contain the index of the task in their filename and the previous restore() was cleaning the files based on this index. When refactoring the BucketingSink we discussed with @aljoscha and we decided to not make the assumption that upon restoring/rescaling new task X will take the state of the previous task X, but states may be re-shuffled arbitrarily. This was also made in order to be future-proof.

Under this new assumption, a new task 0 may take the state of the old task 2. If the new task 2 deletes all the pending files that have its index just because they are no longer in its new state, then task 0 will not be able to recover them.

**Comment by zentol** [comment]
No, it's not just that one method. The method is on a whole other level; it is so easy to edit a commit there simply is no reason for such a thing to exist in the *initial PR*.

So, let's go through a few examples of purely cosmetic changes that could easily have been moved into separate commit. These changes have absolutely no bearing on the core of the PR, which is the port to the new state interface / rescaling.
* Removing `State#hasBucketState`
* Addition, or rather full propagation of getPendingPathFor/etc. in the entire class
* all the random small formatting changes; removing iniitializations to null, generic parameters stuff, removing commas, adding missing spaces
* plenty small comment fixes

Note that these changes are all *good*, it was just be nice to separate these *cosmetic* changes from the *functional* ones.

Now, let's talk about changes that are plain unnecessary and just noise:
* the modifications to reflectTruncate
* the loop change in snapshotState
* no longer storing the subtaskIndex in a field
* no longer looping over all buckets in restoreState (now handleRestoredState)

With a properly cleaned diff i was able to determine very quickly what has actually changed; and this was simply not possible in the current state of the PR.

That is was I'm referring to.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 12]
```diff
@@ -341,23 +355,49 @@ public void setInputType(TypeInformation<?> type, ExecutionConfig executionConfi
...
 	@Override
-	public void open(Configuration parameters) throws Exception {
-		super.open(parameters);
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		Preconditions.checkArgument(this.restoredBucketStates == null, "The operator has already been initialized.");
 
-		subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+		initFileSystem();
 
-		state = new State<T>();
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: undoing it was a bit of an overstatement; moving it into a separate commit with all the other formatting changes would do the trick as well. If you try that out you will see that the changes to open and initializeState no longer intersect so heavily.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 5 above)
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: I don't understand why the subtask index even matters. In a simplified view, every subtask maintains a set of `File`s. When rescaling/restoring state, this set of files is given to another subtask. If this set contains a file marked as pending, it should be allowed to delete it, since no other task should have the same file since the state containing the information about the file (aka, the file path) is not given to multiple subtasks.

**Comment by kl0u** [comment]
The subtask matters for files that were opened but they did not make it into a checkpoint because 
a failure occurred after they were opened but before a checkpoint barrier arrived. These are the ones
that are not cleaned up anymore. The rest, that are part of the checkpoint, they are "cleaned up” or 
rolled back if needed.
 
Probably you may understand more how the cleaning up was happening if you checkout the code 
in the master for the cleanup.

**Comment by zentol** [comment]
That makes sense, thanks for the explanation.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 13]
```diff
@@ -161,9 +161,7 @@
 	// These are initialized with some defaults but are meant to be changeable by the user
 
 	/**
-	 * The default maximum size of part files.
-	 * <p>
-	 * By default, {@code 6 X} the default block size.
+	 * The default maximum size of part files (currently {@code 384 MB}).
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: do we actually know what this value is based on ? :D

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java [hunk 14]
```diff
@@ -69,6 +69,9 @@
 
 	private static String outPath;
 
+	private final String pendingSuffix = ".pending";
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java]: by conventions these should be static; this makes it more obvious to readers that they really never change.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 13 above)
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: Actually this must be based on the old defaults of the HDFS block size. A big file in HDFS was split into smaller of this size and each one was stored on a separate node. If I remember correctly 64MB was the default but they may have changed since then.

**Comment by kl0u** [comment]
Sorry I just amended my previous commit with the comment integration. :S

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java [hunk 15]
```diff
@@ -71,6 +71,10 @@
 	private static org.apache.hadoop.fs.FileSystem dfs;
 	private static String hdfsURI;
 
+	private static final String pendingSuffix = ".pending";
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java]: by convention, the variable name should be upper-case, like `PENDING_SUFFIX`.

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java [hunk 16]
```diff
@@ -118,48 +150,290 @@ public static void destroyHDFS() {
...
+		testHarness2.initializeState(mergedSnapshot);
+		testHarness2.open();
+
+		checkFs(outDir, 0, 0, 5, 5);
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness3 = createRescalingTestSink(outDir, 3, 2, 100);
+		testHarness3.setup();
+		testHarness3.initializeState(mergedSnapshot);
+		testHarness3.open();
+
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java]: let's add a separate check here to make sure that nothing was changed just by the initialization.

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java [hunk 17]
```diff
@@ -70,7 +71,37 @@
...
+				@Override
+				public Path getBucketPath(Clock clock, Path basePath, String element) {
+					return new Path(basePath, element);
+				}
+			})
+			.setWriter(new StringWriter<String>())
+			.setInactiveBucketCheckInterval(inactivityInterval)
+			.setInactiveBucketThreshold(inactivityInterval)
+			.setPartPrefix("part")
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java]: this should also be moved into a field since we use it in checkFS

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultTolerance2ITCase.java [hunk 18]
```diff
@@ -61,6 +62,10 @@
 * high. This provokes the case that the sink restarts without any checkpoint having been performed.
 * This tests the initial cleanup of pending/in-progress files.
 */
+
+// I suggest to remove this as we no longer clean up lingering files.
+
+@Ignore
 public class BucketingSinkFaultTolerance2ITCase extends StreamFaultToleranceTestBase {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultTolerance2ITCase.java]: according to the javadocs we can indeed remove this test. Even without that, i can't quite figure out where the test actually verified that the files were deleted :/

**Comment by zentol** [comment]
Do we provide anything that allows the user to identify files that weren't cleaned up but can be removed? Something like a list of "active" pending files?

**Comment by kl0u** [comment]
Hi @zentol . I integrated your last comments. 

Now for the your last question, the answer is that the only way to distinguish between the two types of files is by their filename (prefix and suffix). This holds for infinite streams. 

Unfortunately in the case of finite streams, we cannot even do that because the current `RichFunction` interface does not allow to distinguish between a failure and normal termination, so the `close()` just leaves the files in `pending` state.

I have opened a discussion in the dev mailing list for that. The thread is this one:
[(http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Adding-a-dispose-method-in-the-RichFunction-td14466.html#a14468)]

Feel free to jump in.

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java [hunk 19]
```diff
@@ -160,7 +167,15 @@ public void postSubmit() throws Exception {
 				while (line != null) {
 					Matcher matcher = messageRegex.matcher(line);
 					if (matcher.matches()) {
-						numRead++;
+						uniqMessagesRead.add(line);
+
+						// check that in the committed files there are no duplicates
+						if (!file.getPath().toString().endsWith(IN_PROGRESS_SUFFIX) && !file.getPath().toString().endsWith(PENDING_SUFFIX)) {
+							if (!messagesInCommittedFiles.add(line)) {
+								Assert.fail("Duplicate entry in committed bucket.");
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java]: this check isn't truly necessary, is it? We have the total number of messages with `readNumbers.size()`, and the number of unique messages ones in `uniqMessagesRead.size()`. Since we compare both with `NUM_STRINGS` the test can only succeed if no duplicates exist anyway.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 20]
```diff
@@ -570,139 +563,165 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
 
 	/**
 	 * Gets the truncate() call using reflection.
-	 *
 	 * <p>
-	 * Note: This code comes from Flume
+	 * <b>NOTE:</b> This code comes from Flume.
 	 */
 	private Method reflectTruncate(FileSystem fs) {
-		Method m = null;
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: we can heavily reduce the changes made here, by putting
```
if (this.refTruncated != null) {
	return this.refTruncated;
}
```
at the start of the method. the remaining code will remain identical, aside from the new assignment near the end.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 21]
```diff
@@ -972,9 +964,13 @@ public void restoreState(State<T> state) {
 	 * <p>
 	 * This should only be disabled if using the sink without checkpoints, to not remove
 	 * the files already in the directory.
+	 *
+	 * <p>
+	 * <b>NOTE: </b> This option is deprecated and remains only for backwards compatibility.
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: this should use the `@deprecated` javadoc annotation

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java [hunk 22]
```diff
@@ -570,139 +563,165 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
...
+				}
+
+				try {
+					fs.delete(testPath, false);
+				} catch (IOException e) {
+					LOG.error("Could not delete truncate test file.", e);
+					throw new RuntimeException("Could not delete truncate test file.", e);
+				}
 			}
+			this.refTruncate = m;
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: I'm not so sure about this assignment; it seems very odd to both assign it to a field and return it. The returned value will typically be assigned to the very same field. The cotnract should be cleaned up to either return the value or assign it.

File: flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java (hunk 19 above)
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFaultToleranceITCase.java]: This test will change after introducing a `dispose()` in the `RichFunction`. The reason for the check is that now when we close and given that we do not delete pending/invalid files upon restore, we cannot check the exactly-once so this test checks the "at-least" once. So I would suggest to leave it as is for now and adapt it as soon as the other change gets in. When this happens, the method that was not used will be the new `close()` and it will rename all valid pending files to committed.

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 21 above)
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: it has it on the next line.

**Comment by kl0u** [comment]
Done @zentol

File: flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java (hunk 21 above)
Comment (zentol) [inline flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java]: no, i mean the actual javadoc annotation, the javadoc should be as follows:
```
/**
	 * Disable cleanup of leftover in-progress/pending files when the sink is opened.
	 *
	 * <p>
	 * This should only be disabled if using the sink without checkpoints, to not remove
	 * the files already in the directory.
	 *
	 * @deprecated This option remains only for backwards compatibility and has no effect.
	 */
```
